File: 代码随想录训练营/day18(112).cpp
//
// Created by StarkLu on 2023/6/10.
//
#include "DataStruct.h"
#include "iostream"
class Solution{
    private:
    bool traversal(TreeNode* root, int target){
        //确定结束条件
        if (root -> right == nullptr && root -> left == nullptr && target == 0) return true;
        if (root -> right == nullptr && root -> left == nullptr && target != 0) return false;
        //单层递归
        if(root -> left){
            target -= root->left -> val;
            if (traversal(root -> left, target)) return true;
            target += root -> left -> val;
        }
        if(root -> right){
            target -= root -> right -> val;
            if (traversal(root -> right, target)) return true;
            target += root -> right -> val;
        }
        return false;
    }
    public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (root == nullptr) return false;
        return traversal(root, targetSum - root -> val);
    }

};

File: 代码随想录训练营/day18(112).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root is None: return False
        return self.helper(root, targetSum - root.val)
    def helper(self, cur, count):
        #决定递归结束条件
        if cur.left is None and cur.right is None and count == 0:return True
        if cur.left is None and cur.right is None and count != 0:return False
        if cur.left:
            count -= cur.left.val
            if self.helper(cur.left, count):
                return True
            count += cur.left.val
        
        if cur.right:
            count -= cur.right.val
            if self.helper(cur.right, count):
                return True
            count += cur.right.val
        return False